### DISCUSSION TITLE
New package: unist-lsp

### DISCUSSION THREAD

**Discussion by remcohaszing** [syntax-tree/unist#80] [Ideas] [answered]
I’ve started working on MDX intellisense today. It’s very raw, getting things up and running first.

Anyway, I need to use functions that already exist in  `unified-language-server`:

```js
/**
 * Convert a unist point to an LSP position.
 *
 * @param {import('unist').Point} point The point to convert
 * @returns {import('vscode-languageserver-types').Position} The converted position
 */
export function unistPointToLspPosition(point) {
  return {
    character: point.column - 1,
    line: point.line - 1,
  }
}

/**
 * Convert a LSP position to a unist point.
 *
 * @param {import('vscode-languageserver-types').Position} position The LSP position to convert.
 * @returns {import('unist').Point} The converted unist point.
 */
export function lspPositionToUnistPoint(position) {
  return {
    column: position.character + 1,
    line: position.line + 1,
  }
}

/**
 * Convert a unist position to an LSP range.
 *
 * @param {import('unist').Position} position The position to convert
 * @returns {import('vscode-languageserver-types').Range} The converted range
 */
export function unistPositionToLspRange(position) {
  return {
    start: unistPointToLspPosition(position.start),
    end: unistPointToLspPosition(position.end),
  }
}

/**
 * Convert a LSP range to a unist position.
 *
 * @param {import('vscode-languageserver-types').Range} range The LSP range to convert.
 * @returns {import('unist').Position} The converted unist position.
 */
export function lspRangeToUnistPosition(range) {
  return {
    start: lspPositionToUnistPoint(range.start),
    end: lspPositionToUnistPoint(range.end),
  }
}
```

It may be small, but I suggest to create a new package for this: `unist-lsp`. I think syntax-tree is the right organization for this. I’m not sure yet if [`vfileMessageToDiagnostic`](https://github.com/unifiedjs/unified-language-server/blob/main/lib/index.js#L108-L136) should be a part of it too.

**Comment by wooorm** [comment]
Sounds like two separate packages? Good idea otherwise! Need help?

**Reply by remcohaszing** [reply]
How do you envision this to be 2 separate packages? I could see this in multiple ways.

Since I’m new to creating new repositories in unified, are there best practices for this? I.e. just create the repo and push to main? Or create an empty repo and create a PR for review?

**Reply by wooorm** [reply]
> How do you envision this to be 2 separate packages?

One in `syntax-tree`, one in `vfile` (my comment was related to you `vfileMessageToDiagnostic should be a part of it too.`

>  I.e. just create the repo and push to main? Or create an empty repo and create a PR for review?

I clone a similar, recent example. `rm -rf .git`. Update everything accordingly.
I typically push everything when it works as an `initial commit` / `.`.
If everything passes, I cut `1.0.0`.

You can:
a) push a small portion at the start, and everything you want reviews on separately as PRs
b) push everything when it works, open an issue to get reviews if needed?

**Comment by remcohaszing** [comment, marked as answer]
This has been released as [`unist-util-lsp`](https://github.com/syntax-tree/unist-util-lsp).
